### PR TITLE
PR #9714: JVM arguments and options

### PR DESCRIPTION
**Pull request by vadeg**
PR #9714 done.
* Added `systemPropertyVariables` configuration tag support for JVM system properties. (like [maven-surefire-plugin](http://maven.apache.org/components/surefire/maven-surefire-plugin/examples/system-properties.html) does).
* Added several unit and integration tests.

Example how to configure `jvmOptions` and `systemPropetries`
```
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <version>${some-version}</version>
    <executions>
        <execution>
            <configuration>
                <jvmArguments>-Dfoo="value 1" -Dbar=value2</jvmArguments>
                <systemPropertyVariables>
                    <property1>value1</property1>
                    <property2 />
                </systemPropertyVariables>
            </configuration>
        </execution>
    </executions>
</plugin>
```

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@vadeg Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=10741)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=10741) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@vadeg Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=10741)!

**Comment by vadeg** [comment]
Are there any chances to see why build is failing? Unfortunately, I do not have permissions to CI.

**Comment by snicoll** [comment]
See #10693 

There are checkstyle errors with your change. Building the project on the command line (`mvn package`) should reveal those errors. It's something you should do before submitting a PR IMO

```
[INFO] There are 18 errors reported by Checkstyle 7.6 with src/checkstyle/checkstyle.xml ruleset.
[ERROR] src/main/java/org/springframework/boot/maven/RunArguments.java:[21,8] (imports) UnusedImports: Unused import - java.util.List.
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[16] (header) RegexpHeader: Line does not match expected header line of '^$'.
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[34,24] (coding) RequireThis: Reference to instance variable 'key' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[43,24] (coding) RequireThis: Reference to instance variable 'value' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[52,21] (coding) RequireThis: Reference to instance variable 'key' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[55,21] (coding) RequireThis: Reference to instance variable 'value' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[55,38] (coding) RequireThis: Reference to instance variable 'value' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[56,63] (coding) RequireThis: Reference to instance variable 'key' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[58,56] (coding) RequireThis: Reference to instance variable 'key' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/SystemProperty.java:[58,61] (coding) RequireThis: Reference to instance variable 'value' needs "this.".
[ERROR] src/main/java/org/springframework/boot/maven/AbstractRunMojo.java:[27] (imports) ImportOrder: 'org.springframework.boot.loader.tools.FileUtils' should be separated from previous imports.
[ERROR] src/main/java/org/springframework/boot/maven/AbstractRunMojo.java:[30] (imports) ImportOrder: Wrong order for 'java.io.File' import.
[ERROR] src/main/java/org/springframework/boot/maven/AbstractRunMojo.java:[35] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - java.util.*.
[ERROR] src/main/java/org/springframework/boot/maven/AbstractRunMojo.java:[171] (javadoc) JavadocStyle: First sentence should end with a period.
[ERROR] src/main/java/org/springframework/boot/maven/AbstractRunMojo.java:[333,66] (coding) RequireThis: Reference to instance variable 'jvmOptions' needs "this.".
[ERROR] src/test/java/org/springframework/boot/maven/SystemPropertiesTests.java:[1] (header) RegexpHeader: Line does not match expected header line of '^\Q/*\E$'.
[ERROR] src/test/java/org/springframework/boot/maven/JvmOptionsTests.java:[1] (header) RegexpHeader: Line does not match expected header line of '^\Q/*\E$'.
[ERROR] src/test/java/org/springframework/boot/maven/JvmOptionsTests.java:[7] (imports) ImportOrder: Wrong order for 'java.io.File' import.
[INFO] ------------------------------------------------------------------------
```

**Comment by vadeg** [comment]
@snicoll Thank you! I will fix it. If run locally there are two test failures. 
One is related to slow application context startup. 
Second one is related to slow netty startup time. By default it waits 3 seconds to start [see react-netty PR #184](https://github.com/reactor/reactor-netty/pull/184).

File: spring-boot-project/spring-boot-parent/pom.xml [hunk 1]
```diff
@@ -22,8 +22,8 @@
 		<java.version>1.8</java.version>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-		<aether.version>1.0.2.v20150114</aether.version>
-		<maven.version>3.1.1</maven.version>
+		<aether.version>1.1.0</aether.version>
+		<maven.version>3.2.5</maven.version>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-parent/pom.xml]: Please limit the scope of your PR. Those dependencies upgrade shouldn't be part of this commit. If they are necessary please explain why

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/pom.xml [hunk 2]
```diff
@@ -47,6 +47,7 @@
 		<dependency>
 			<groupId>org.apache.maven.shared</groupId>
 			<artifactId>maven-common-artifact-filters</artifactId>
+			<version>3.0.1</version>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/pom.xml]: Ditto

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Class for system properties passed from pom file.
+ * @author Vadim Shaigorodskiy
+ * @see RunMojo
+ * @see StartMojo
+ */
+public class SystemProperty {
+
+	private final static String NO_VALUE_FORMAT = "-D%s";
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java]: We use `public static final`

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml [hunk 4]
```diff
@@ -0,0 +1,821 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!-- Generated by maven-plugin-tools 3.5 on 2017-10-21 -->
+
+<plugin>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml]: I am not keen to commit something that  was auto-generated. Can you please explain why you've done this?

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/pom_examples/jvm-args-1.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/pom_examples/jvm-args-1.xml]: We have integration tests for this. This sample project should move to `src/test/it`

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	private final static String KEY_VALUE_FORMAT = NO_VALUE_FORMAT + "=%s";
+
+	private String key;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java]: Whatever we do, I see no reason to do any differently from what core maven plugins do. The surefire plugin had several rounds of trials apparently so I'd go with [whatever their last version is doing](http://maven.apache.org/components/surefire/maven-surefire-plugin/examples/system-properties.html)

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/AbstractRunMojo.java [hunk 7]
```diff
@@ -169,6 +173,22 @@
 	@Parameter(property = "spring-boot.run.skip", defaultValue = "false")
 	private boolean skip;
 
+	/**
+	 * List of JVM options.
+	 * @since 2.0
+	 */
+	@Parameter(property = "spring-boot.run.jvmOptions")
+	private List<String> jvmOptions;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/AbstractRunMojo.java]: I am not entirely convinced that this is required. This looks quite verbose with not a lot of benefit. Surefire has an [argLine parameter](http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#argLine) that matches what `jvmArguments` does so I'd keep that (no need for deprecation).

File: spring-boot-project/spring-boot-parent/pom.xml (hunk 1 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-parent/pom.xml]: Maven documentation [recommends](https://maven.apache.org/plugin-developers/plugin-testing.html) to use `maven-plugin-testing-harness` library as a simple way for creating integration tests for Mojo's. `maven-plugin-testing-harness` support maven starting from version `3.2.5`. Updating maven's version triggers updates of dependent libraries. 

I agree that this is should be another PR. I tried to make minimal set of version changes in this  PR related to maven. Actually I was thinking about creating a separate PR with maven 3.5.0 support and clean up maven-related library "hell" there but I am not sure maybe you have strong reasons to support maven from version 3.2.5. If you need it I can create a separate ticket/PR for this and postpone current PR.

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/pom.xml (hunk 2 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/pom.xml]: answered above.

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java (hunk 3 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java]: ok

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/AbstractRunMojo.java (hunk 7 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/AbstractRunMojo.java]: ok

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java (hunk 6 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/main/java/org/springframework/boot/maven/SystemProperty.java]: There are a lot of core maven plugins and far not all of them doing configuration like surefire plugin does. 
For example `maven-invoker-plugin` allows to [configure properties](https://maven.apache.org/plugins/maven-invoker-plugin/run-mojo.html#properties) in two ways: like a string line and like a map.

`maven-compiler-plugin` allows to configure [compiler arguments](https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#compilerArgs) like I did with `<jvmOptions>`

Why it should be done exactly like in `maven-surefire-plugin`? To be clear I am not against that. I have just did it in one of the ways it was discussed in PR. If the `maven-surefire-plugin` is the source of truth then I can make it exactly like there.

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml (hunk 4 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml]: It is needed for mojo integration tests. By default `maven-plugin-testing-harness` tool is using plugin descriptor from `META-INF/maven/plugin.xml` which is generated automatically. The problem is that generated file will have resolved plugin version (like `2.0.0-SNAPSHOT` or similar) but the tool is using current project version to for searching plugin definition in test `pom.xml` files. In current state this version will be `${revision}`. In short it needs that `version` tag value in `plugin.xml` was the same as project.version in project's `pom.xml`. Thats why I have generated this and put in tests.

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/pom_examples/jvm-args-1.xml (hunk 5 above)
Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/pom_examples/jvm-args-1.xml]: You mean `spring-boot-maven-plugin/src/it`?

Comment (snicoll) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/pom_examples/jvm-args-1.xml]: Yes

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml (hunk 4 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml]: OK. Sorry but I am not keen to merge that. Can you write an IT instead?

Comment (vadeg) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/test/resources/plugin-descriptor.xml]: Sure. But before that can we agree on xml format for properties?

**Comment by vadeg** [comment]
@snicoll I have updated PR based on your comments. Could you please review again?

**Comment by vadeg** [comment]
Any updates on this?

**Comment by heruan** [comment]
I'm very interested on this too, currently if I set a property via `<jvmArguments>` in the plugin config and then I use the same property in the command line to add other options, the ones defined in the config are not applied.

**Comment by snicoll** [comment]
> I use the same property in the command line to add other options, the ones defined in the config are not applied.

That PR isn't going to change that. That's just how Maven works.

**Comment by heruan** [comment]
If I have:

```xml
<systemPropertyVariables>
    <foo>bar</foo>
</systemPropertyVariables>
```

and then I run `mvn spring-boot:run -Dbaz=qux` wouldn't also `-Dfoo=bar` be applied?

**Comment by snicoll** [comment]
No, `-Danything=value` is the standard way to  pass a system property to a process. The very reason why the plugin allows to be configured for such thing is because it can fork a process, making that completely irrelevant. If you have more questions, please ask on Gitter.

**Comment by snicoll** [comment]
@vadeg thanks a lot for your first contribution to Spring Boot. This is now merged in `master`.

**Comment by snicoll** [comment]
If the value contains space, it isn't properly wrapped, I forgot to review that.

**Comment by vadeg** [comment]
@snicoll should I update PR with this fix?

**Comment by snicoll** [comment]
No need. I’ve fixed that myself already. Thanks!
